tools/git-hooks/commit-msg

```diff
@@ -0,0 +1,339 @@
+    pointed        points        pointing
+    pruned         prunes        pruning
+    ran            runs          running
+    refactored     refactors     refactoring
+    released       releases      releasing
+    removed        removes       removing
+    renamed        renames       renaming
+    replaced       replaces      replacing
+    resolved       resolves      resolving
+    reverted       reverts       reverting
+                   sets          setting
+    showed         shows         showing
+    swapped        swaps         swapping
+    tested         tests         testing
+    tidied         tidies        tidying
+    updated        updates       updating
+    upped          ups           upping
+    used           uses          using
+  )
+
+  # enable case insensitive match
+  shopt -s nocasematch
+
+  for BLACKLISTED_WORD in "${IMPERATIVE_MOOD_BLACKLIST[@]}"; do
+    [[ ${COMMIT_SUBJECT} =~ ^[[:blank:]]*$BLACKLISTED_WORD ]]
+    test $? -eq 0 && add_warning 1 "Use the imperative mood in the subject line, e.g 'fix' not 'fixes'" && break
+  done
+
+  # disable case insensitive match
+  shopt -u nocasematch
+
+  # 6. Wrap the body at 72 characters
+  # ------------------------------------------------------------------------------
+
+  URL_REGEX='^[[:blank:]]*(https?|ftp|file|wss?|git|ssh|data|irc|dat)://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]'
+
+  for i in "${!COMMIT_MSG_LINES[@]}"; do
+    LINE_NUMBER=$((i+1))
+    test "${#COMMIT_MSG_LINES[$i]}" -le 72 || [[ ${COMMIT_MSG_LINES[$i]} =~ $URL_REGEX ]]
+    test $? -eq 0 || add_warning $LINE_NUMBER "Wrap the body at 72 characters (${#COMMIT_MSG_LINES[$i]} chars)"
+  done
+
+  # 7. Use the body to explain what and why vs. how
+  # ------------------------------------------------------------------------------
+
+  # ?
+
+  # 8. Do no write single worded commits
+  # ------------------------------------------------------------------------------
+
+  COMMIT_SUBJECT_WORDS=(${COMMIT_SUBJECT})
+  test "${#COMMIT_SUBJECT_WORDS[@]}" -gt 1
+  test $? -eq 0 || add_warning 1 "Do no write single worded commits"
+
+  # 9. Do not start the subject line with whitespace
+  # ------------------------------------------------------------------------------
+
+  [[ ${COMMIT_SUBJECT} =~ ^[[:blank:]]+ ]]
+  test $? -eq 1 || add_warning 1 "Do not start the subject line with whitespace"
+}
+
+#
+# It's showtime.
+#
+
+set_colors
+
+set_editor
+
+if tty >/dev/null 2>&1; then
+  TTY=$(tty)
+else
+  TTY=/dev/tty
+fi
+
+while true; do
+
+  read_commit_message
+
+  validate_commit_message
+
+  # if there are no WARNINGS are empty then we're good to break out of here
+  test ${#WARNINGS[@]} -eq 0 && exit 0;
+
+  display_warnings
+
+  # if non-interactive don't prompt and exit with an error
+  if [ ! -t 1 ] && [ -z ${FAKE_TTY+x} ]; then
+    exit 1
+  fi
+
+  # Ask the question (not using "read -p" as it uses stderr not stdout)
+  echo -en "${BLUE}Proceed with commit? [e/y/n/?] ${NC}"
```

Comment:
I wondered a bit what `e` means, then I realized it must be edit. Do you have an idea how this could be made more clear? Maybe similar to git when resolving merge conflicts
```
echo -en "${BLUE}Proceed with commit?"
echo -en "${BLUE}(y)es, (n)o, (e)dit: ${NC}"
```
What do you think?

Reply:
There is a help text when you enter `?` but yes we can make it more explicit.

Reply:
Oh, one can enter `?`? I didn't assume that :smile: